### DISCUSSION TITLE
Is the only purpose of `extend_ref` to retrieve object signer?

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#3] [Questions] [answered]
I see we do this pattern a lot when using an object as the owner of an NFT collection. Then later we generate the NFT collection's `signer` from `extend_ref` in order to mint NFT.

```rust
fun init_module(account: &signer) {
  let constructor_ref = object::create_named_object(
      account,
      OBJECT_SEED,
  );
  let extend_ref = object::generate_extend_ref(&constructor_ref);
  let object_signer = &object::generate_signer(&constructor_ref);
  move_to(object_signer, CollectionCapability {
      extend_ref,
  });
  collection::create_unlimited_collection(
      object_signer,
      description,
      name,
      option::none(),
      uri,
  )
}

public entry fun mint_aptogotchi_nft(user: &signer) {
  let collection_address = get_collection_address();
  let constructor_ref = &token::create(
      &get_collection_signer(collection_address),
      utf8(APTOGOTCHI_COLLECTION_NAME),
      description,
      token_name,
      option::none(),
      uri,
  );
  let token_signer_ref = &object::generate_signer(constructor_ref);
  let aptogotchi = Aptogotchi {
      parts,
  };
  move_to(token_signer_ref, aptogotchi);
}
```

Code pulled from [aptogotchi repo](https://github.com/aptos-labs/aptogotchi/).

**Comment by 0x-j** [comment]
Sample answer.

**Comment by gregnazario** [comment, marked as answer]
Yes, an extend ref is specifically to generate the object's signer at a later time.

`signer` purposely cannot be stored, as it's an authorization object for an address.  Instead, an `ExtendRef` can be stored to generate this signer to either `move_to` new items, or to authorize operations on an address like such as `TokenObjects`. for example `aptos_token_objects::token::create()`

**Reply by JohnChangUK** [reply]
I would say almost all the cases it would be to generate the `signer` as Greg said. There are a couple of instances where we use the `ExtendRef` to "extend" or "upgrade" the object, as we see in collection.move and fungible_asset.move in the function:

```
public fun upgrade_to_concurrent(
  ref: &ExtendRef,
)
```

[Source](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/fungible_asset.move#L633-L635)

**Comment by dbrdz** [comment]
What is the purpose of creating a collection using the Object's signer? What is the difference between doing this vs creating a resource account and creating a collection using the resource account's signer?

**Reply by 0x-j** [reply]
Object is basically the successor to resource account, you can do everything you did with resource account using object now, with better performance.
